zulipterminal/ui_tools/boxes.py

```diff
@@ -636,10 +636,29 @@ def autocomplete_streams(
 
         prefix_length = len(prefix_string)
 
-        matched_data = match_stream(
+        _, matched_streams = match_stream(
             stream_data, text[prefix_length:], self.view.pinned_streams
         )
-        return matched_data
+
+        muted_stream_names = [
+            stream_name
+            for stream_name in matched_streams
+            if stream_name
+            in [
+                self.model.stream_dict[stream_id]["name"]
+                for stream_id in self.model.muted_streams
+            ]
+        ]
+
+        for muted_stream in muted_stream_names:
+            matched_streams.remove(muted_stream)
+            if muted_stream in [stream["name"] for stream in self.view.pinned_streams]:
+                muted_stream_names.remove(muted_stream)
+                muted_stream_names.insert(0, muted_stream)
+        matched_streams.extend(muted_stream_names)
+
+        matched_stream_typeaheads = format_string(matched_streams, "#**{}**")
+        return matched_stream_typeaheads, matched_streams
```

Comment:
I haven't followed this PR a lot but I feel that all of this code should belong inside `match_stream`. Is there any reason for not doing so? We can send in the model variables as arguments into the function and process everything there.

Reply:
@neiljp On a design level, I wanted to ask why, in the model, we treat `muted_streams` different from `(un)pinned_streams`, i.e., different data types.

Reply:
And about data type of ``muted_streams``, I agree with using ``stream_names`` instead of ``stream_ids`` like we have implemented for ``(un)pinned_streams``

Reply:
Oh I see, but a boolean like `prioritize_unmuted` might help? Since `match_stream` also does some sorting apart from matching, it's better to place all sorting code together.
* If False, it'll return pinned matches then unpinned matches.
* If True, it'll return pinned unmuted matches, unpinned unmuted matches, pinned muted matches then unpinned muted matches.

Reply:
@Rohitth007 I agree there are multiple ways of writing the code, and we may even update the stream order later - or add options - but for now I'm happy for this to be focused on the functionality and ensuring this behaves as expected by using the new tests (and test structure). Once we have tests, we can refactor more safely. I think @srdeotarse already rewrote the code, and we still may need to adjust the algorithm.

Reply:
Re muted_streams vs (un)pinned_streams, these are data-structures used for different purposes right now. The former is simply to know if a stream is muted, while the latter is designed for the UI (historically).

Regarding the tests, I would favor having names in the test cases to make them more readable, independently of the data in the model. That is eg. `{"pinned": {"Stream 1"}, "muted": {"Secret Stream"}}` is more readable than the latter being an int.

Reply:
The reason for having it inside `match_stream` is it would simplify the logic by a lot. Here we are throwing away the matched typeaheads, resorting an already sorted list given by `match_streams`  and then recreating new matched typeaheads. If this is done inside `match_stream` then it would barely take 5-6 lines and it wouldn't affect any of the tests.

Another reason is we would barely have to do any work to extend this feature to the stream panel.